File: app/src/androidTest/java/com/protonvpn/actions/ProfilesRobot.java
/*
 * Copyright (c) 2018 Proton Technologies AG
 *
 * This file is part of ProtonVPN.
 *
 * ProtonVPN is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * ProtonVPN is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with ProtonVPN.  If not, see <https://www.gnu.org/licenses/>.
 */
package com.protonvpn.actions;

import com.protonvpn.MockSwitch;
import com.protonvpn.android.R;
import com.protonvpn.android.models.config.TransmissionProtocol;
import com.protonvpn.android.models.config.VpnProtocol;
import com.protonvpn.android.models.vpn.VpnCountry;
import com.protonvpn.results.ConnectionResult;
import com.protonvpn.results.ProfilesResult;
import com.protonvpn.testsHelper.ConditionalActionsHelper;
import com.protonvpn.testsHelper.UIActionsTestHelper;

public class ProfilesRobot extends UIActionsTestHelper {

    public ServiceRobot serviceRobot;
    public ProfilesResult profilesResult;

    public ProfilesRobot() {
        serviceRobot = new ServiceRobot();
        profilesResult = new ProfilesResult();
    }

    public ProfilesRobot selectColor(final int resourceId) {
        clickOnObjectWithContentDescription(resourceId);
        return this;
    }

    public ProfilesRobot clickOnFastestOption() {
        clickOnObjectWithText("Fastest");
        return this;
    }

    public ConnectionResult clickOnConnectButton(String profileName) {
        clickOnObjectWithIdAndContentDescription(R.id.buttonConnect, profileName);
        if (!MockSwitch.mockedConnectionUsed) {
            new HomeRobot().allowToUseVpn();
        }
        return new ConnectionResult();
    }

    public ProfilesRobot clickOnCreateNewProfileButton() {
        clickOnObjectWithText("CREATE NEW PROFILE");
        return this;
    }

    public ProfilesRobot insertTextInProfileNameField(String text) {
        insertTextIntoFieldWithId(R.id.editName, text);
        return this;
    }

    public ProfilesRobot selectFirstCountry() {
        clickOnObjectWithId(R.id.spinnerCountry);
        clickOnObjectWithText(serviceRobot.getFirstCountryFromBackend());
        return this;
    }

    public ProfilesRobot selectFirstSecureCoreExitCountry() {
        clickOnObjectWithId(R.id.spinnerCountry);
        clickOnObjectWithText(serviceRobot.getFirstSecureCoreExitCountryFromBackend().getCountryName());
        return this;
    }

    public ProfilesRobot selectSecondSecureCoreExitCountry() {
        clickOnObjectWithId(R.id.spinnerCountry);
        clickOnObjectWithText(serviceRobot.getSecondSecureCoreExitCountryFromBackend().getCountryName());
        return this;
    }

    public ProfilesRobot selectSecureCoreEntryCountryForSecondExit() {
        VpnCountry exitCountry = serviceRobot.getSecondSecureCoreExitCountryFromBackend();
        clickOnObjectWithId(R.id.spinnerServer);
        clickOnObjectWithText(serviceRobot.getSecureCoreEntryCountryFromBackend(exitCountry));
        return this;
    }

    public ProfilesRobot selectFirstNotAccessibleVpnCountry() {
        clickOnObjectWithId(R.id.spinnerCountry);
        clickOnObjectWithText(serviceRobot.getFirstNotAccessibleVpnCountryFromBackend() + " (Upgrade)");
        return this;
    }

    public ProfilesRobot selectRandomServer() {
        clickOnObjectWithId(R.id.spinnerServer);
        // TODO Use "Random" instead of Fastest once random profile problems are solved
        clickOnObjectWithText("Fastest");
        return this;
    }

    public ProfilesResult selectNonAccessibleRandomServer() {
        clickOnObjectWithId(R.id.spinnerServer);
        clickOnObjectWithText("Random (Upgrade)");
        return new ProfilesResult();
    }

    public ProfilesResult clickOnSaveButton() {
        clickOnObjectWithId(R.id.fabSave);
        return new ProfilesResult();
    }

    public ProfilesRobot selectProfile(String profileName) {
        clickOnObjectWithText(profileName);
        return this;
    }

    public ProfilesResult navigateBackFromForm() {
        pressDeviceBackButton();
        return new ProfilesResult();
    }

    public ProfilesResult clickCancelButton() {
        clickOnObjectWithText(R.string.cancel);
        return new ProfilesResult();
    }

    public ProfilesResult clickDiscardButton() {
        clickOnObjectWithText("Discard");
        return new ProfilesResult();
    }

    public ProfilesRobot enableSecureCore() {
        clickOnObjectWithText("Secure Core");
        return this;
    }

    public ConnectionResult clickYesButton() {
        clickOnObjectWithId(R.id.md_buttonDefaultPositive);
        return new ConnectionResult();
    }

    public ConnectionResult clickOnUpgradeButton(String contentDescription) {
        clickOnObjectWithIdAndContentDescription(R.id.buttonUpgrade, contentDescription);
        return new ConnectionResult();
    }

    public ProfilesRobot clickEditProfile() {
        clickOnObjectWithId(R.id.profile_edit_button);
        return this;
    }

    public ProfilesRobot updateProfileName(String newProfileName) {
        insertTextIntoFieldWithId(R.id.editName, newProfileName);
        return this;
    }

    public ProfilesRobot clickDeleteProfile() {
        clickOnObjectWithId(R.id.action_delete);
        clickOnObjectWithText("Delete");
        return this;
    }

    public ProfilesRobot selectOpenVPNProtocol() {
        ConditionalActionsHelper.scrollDownInViewWithIdUntilObjectWithIdAppears(R.id.coordinator,
            R.id.spinnerDefaultProtocol);
        clickOnObjectWithId(R.id.spinnerDefaultProtocol);
        clickOnObjectWithText(VpnProtocol.OpenVPN.toString());
        return this;
    }

    public ProfilesRobot selectUDPTransmissionProtocol() {
        ConditionalActionsHelper.scrollDownInViewWithIdUntilObjectWithIdAppears(R.id.coordinator,
            R.id.spinnerTransmissionProtocol);
        clickOnObjectWithId(R.id.spinnerTransmissionProtocol);
        clickOnObjectWithText(TransmissionProtocol.UDP.toString());
        return this;
    }
}

File: app/src/androidTest/java/com/protonvpn/actions/ProfilesRobot.kt
/*
 * Copyright (c) 2018 Proton Technologies AG
 *
 * This file is part of ProtonVPN.
 *
 * ProtonVPN is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * ProtonVPN is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with ProtonVPN.  If not, see <https://www.gnu.org/licenses/>.
 */
package com.protonvpn.actions

import android.view.View
import androidx.annotation.IdRes
import androidx.test.espresso.Espresso
import androidx.test.espresso.assertion.ViewAssertions
import androidx.test.espresso.matcher.ViewMatchers
import com.protonvpn.android.R
import com.protonvpn.base.BaseRobot
import com.protonvpn.base.BaseVerify
import com.protonvpn.matchers.ProtonMatcher
import com.protonvpn.testsHelper.ConditionalActionsHelper
import me.proton.core.presentation.ui.view.ProtonAutoCompleteInput
import me.proton.core.presentation.ui.view.ProtonCheckbox
import me.proton.core.presentation.ui.view.ProtonInput
import org.hamcrest.Matcher

/**
 * [ProfilesRobot] Contains all actions and verifications for profiles screen
 */
class ProfilesRobot : BaseRobot() {

    private val serviceRobot = ServiceRobot()

    fun clickOnSaveButton(): ProfilesRobot = clickElementById(R.id.action_save)

    fun navigateBackFromForm(): ProfilesRobot = back()

    fun clickCancelButton(): ProfilesRobot = clickElementByText(R.string.cancel)

    fun clickDiscardButton(): ProfilesRobot = clickElementByText(R.string.discard)

    fun clickScSpeedInfoDialogActivate(): ProfilesRobot =
        clickElementByText(R.string.secureCoreActivateDialogButton)

    fun selectColorIndex(index: Int): ProfilesRobot =
        clickElementByIndexInParent(R.id.layoutPalette, index)

    fun clickOnConnectButton(profileName: String): ConnectionRobot =
        clickElementByIdAndContentDescription(R.id.buttonConnect, profileName)

    fun insertTextInProfileNameField(text: String): ProfilesRobot {
        scrollTo(R.id.inputName, ProtonInput::class.java)
        return replaceText(R.id.inputName, text)
    }

    fun selectFirstCountry(): ProfilesRobot {
        scrollToAndClickDropDown(R.id.inputCountry)
        return clickElementByText(serviceRobot.firstCountryFromBackend)
    }

    fun selectSecondSecureCoreExitCountry(): ProfilesRobot {
        scrollToAndClickDropDown(R.id.inputCountry)
        return clickElementByText(serviceRobot.secondSecureCoreExitCountryFromBackend.countryName)
    }

    fun selectSecureCoreEntryCountryForSecondExit(): ProfilesRobot {
        val exitCountry = serviceRobot.secondSecureCoreExitCountryFromBackend
        scrollToAndClickDropDown(R.id.inputServer)
        return clickElementByText(serviceRobot.getSecureCoreEntryCountryFromBackend(exitCountry))
    }

    fun selectRandomServer(): ProfilesRobot {
        scrollToAndClickDropDown(R.id.inputServer)
        return clickElementByText(R.string.profileRandom)
    }

    fun enableSecureCore(): ProfilesRobot {
        ConditionalActionsHelper().scrollDownInViewWithIdUntilObjectWithIdAppears(
            R.id.coordinator,
            R.id.checkboxSecureCore
        )
        return clickElement(R.id.checkboxSecureCore, ProtonCheckbox::class.java)
    }

    fun clickOnUpgradeButton(contentDescription: String): ConnectionRobot {
        view.waitForCondition {
            clickElementByIdAndContentDescription<Any>(R.id.buttonUpgrade, contentDescription)
            Espresso.onView(ViewMatchers.withText(R.string.upgrade_secure_core_title))
                .check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
        }
        return ConnectionRobot()
    }

    fun clickOnConnectButtonUntilConnected(profileName: String): ConnectionRobot {
        view.waitForCondition {
            clickElementByIdAndContentDescription<Any>(R.id.buttonConnect, profileName)
            Espresso.onView(ViewMatchers.withId(R.id.buttonDisconnect))
                .check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
        }
        return ConnectionRobot()
    }

    fun clickEditProfile(): ProfilesRobot {
        view
            .withVisibility(ViewMatchers.Visibility.VISIBLE)
            .withId(R.id.profile_edit_button)
            .click()
        return this
    }

    fun updateProfileName(newProfileName: String): ProfilesRobot =
        replaceText(R.id.inputName, newProfileName)

    fun clickDeleteProfile(): ProfilesRobot {
        ConditionalActionsHelper().scrollDownInViewWithIdUntilObjectWithIdAppears(
            R.id.coordinator,
            R.id.buttonDelete
        )
        clickElementById<ProfilesRobot>(R.id.buttonDelete)
        return clickElementByText(R.string.delete)
    }

    fun clickOnCreateNewProfileButton(): ProfilesRobot {
        clickOnCreateNewProfileUntilMatcherIsDisplayed(ViewMatchers.withId(R.id.action_save))
        return this
    }

    fun clickOnCreateNewProfileUntilUpsellIsShown() : UpsellModalRobot {
        clickOnCreateNewProfileUntilMatcherIsDisplayed(ProtonMatcher.withHtmlText(R.string.upgrade_profiles_text))
        return UpsellModalRobot()
    }

    fun selectOpenVPNProtocol(udp: Boolean): ProfilesRobot {
        scrollToAndClickDropDown(R.id.inputProtocol)
        val protocol =
            if (udp) R.string.settingsProtocolNameOpenVpnUdp
            else R.string.settingsProtocolNameOpenVpnTcp
        return clickElementByText(protocol)
    }

    private fun clickOnCreateNewProfileUntilMatcherIsDisplayed(matcher: Matcher<View>) {
        waitUntilDisplayedByText<Any>(R.string.create_new_profile)
        view.waitForCondition {
            clickElementByText<Any>(R.string.create_new_profile)
            Espresso.onView(matcher)
                .check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
        }
    }

    private fun scrollToAndClickDropDown(@IdRes id: Int): ProfilesRobot {
        ConditionalActionsHelper().scrollDownInViewWithIdUntilObjectWithIdAppears(
            R.id.coordinator,
            R.id.inputProtocol,
            ProtonAutoCompleteInput::class.java
        )
        return clickElement(id, ProtonAutoCompleteInput::class.java)
    }

    private fun scrollTo(@IdRes id: Int, clazz: Class<out Any>): ProfilesRobot {
        ConditionalActionsHelper().scrollDownInViewWithIdUntilObjectWithIdAppears(
            R.id.coordinator,
            id,
            clazz
        )
        return this
    }

    class Verify : BaseVerify() {

        fun errorEmptyNameIsVisible() = checkIfElementIsDisplayedByStringId(R.string.errorEmptyName)

        fun errorEmptyCountryIsVisible() =
            checkIfElementIsDisplayedByStringId(R.string.errorEmptyCountry)

        fun upgradeButtonIsDisplayed() = checkIfElementIsDisplayedById(R.id.buttonUpgrade)

        fun defaultProfileOptionsAreVisible() {
            checkIfElementIsDisplayedByStringId(R.string.profileFastest)
            checkIfElementIsDisplayedByStringId(R.string.profileRandom)
        }

        fun discardChangesDialogIsVisible() =
            checkIfElementIsDisplayedByStringId(R.string.discardChanges)

        fun profileIsNotVisible(profileName: String) {
            view.withId(R.id.textServer).withText(profileName).checkDoesNotExist()
        }

        fun profileIsVisible(profileName: String) {
            view.withId(R.id.textServer).withText(profileName).checkDisplayed()
        }

        fun defaultProfilesArePaid() {
            view.withId(R.id.buttonUpgrade).withContentDesc(R.string.profileFastest).checkDisplayed()
            view.withId(R.id.buttonUpgrade).withContentDesc(R.string.profileRandom).checkDisplayed()
        }
    }

    inline fun verify(block: Verify.() -> Unit) = Verify().apply(block)
}
